<a href="https://colab.research.google.com/github/catalinagaviriamet-spec/Ciencia-de-datos/blob/main/Ejercicio_DataAdults.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Ejercicio Datos Sociodemográficos de adultos

Primera entrega para el módulo Introducción a la Ciencia de Datos

Integrantes:
Yerlith Judith Zabala Martinez,
Carolina Tirado Osorio,
Ana Catalina Gaviria Metaute.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/adult-all.csv"
import pandas as pd
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/adult-all.csv"

columnas_es = [
    "edad",
    "tipo_trabajo",
    "peso_muestral",
    "nivel_educativo",
    "anios_educacion",
    "estado_civil",
    "ocupacion",
    "relacion_familiar",
    "raza",
    "sexo",
    "ganancia_capital",
    "perdida_capital",
    "horas_por_semana",
    "pais_origen",
    "ingresos"
]

df = pd.read_csv(url, header=None, names=columnas_es)
df.head()

print('Shape inicial:', df.shape)
display(df.head())

Shape inicial: (48842, 15)


,edad,tipo_trabajo,peso_muestral,nivel_educativo,anios_educacion,estado_civil,ocupacion,relacion_familiar,raza,sexo,ganancia_capital,perdida_capital,horas_por_semana,pais_origen,ingresos
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


Se muestra que la base de datos elegida contiene un total de 48,842 registros con 15 columnas.

**Exploración Inicial del dataset**: estructura, estadística y valores faltantes

In [2]:
df.describe()

,edad,peso_muestral,anios_educacion,ganancia_capital,perdida_capital,horas_por_semana
count,48842.000000,4.884200e+04,48842.000000,48842.000000,48842.000000,48842.000000
mean,38.643585,1.896641e+05,10.078089,1079.067626,87.502314,40.422382
std,13.710510,1.056040e+05,2.570973,7452.019058,403.004552,12.391444
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.175505e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.781445e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.376420e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.490400e+06,16.000000,99999.000000,4356.000000,99.000000


Del resumen numérico se observa lo siguiente:
1. La edad promedio de las persona es de 38 años, y en promedio han cursado 10 años de educación.
2. La edad mínima de la población de de 17 años y la máxima es de 90.
3. En promedio las personas trabajan 40 horas a la semana, y el máximo de horas trabajadas son 99.
4. Comparando los mínimos y los máximos contra los cuartiles, así como la desviación estándar respecto a la media, se identifican valores outliers en las columnas edad, ganancia_capital, perdida_capital y horas_por_semana, debido a que los máximos están muy alejados del q3 y la std para algunas viriables está muy alta respecto a la media.

In [3]:
df.info()
df.describe()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   edad               48842 non-null  int64 
 1   tipo_trabajo       48842 non-null  object
 2   peso_muestral      48842 non-null  int64 
 3   nivel_educativo    48842 non-null  object
 4   anios_educacion    48842 non-null  int64 
 5   estado_civil       48842 non-null  object
 6   ocupacion          48842 non-null  object
 7   relacion_familiar  48842 non-null  object
 8   raza               48842 non-null  object
 9   sexo               48842 non-null  object
 10  ganancia_capital   48842 non-null  int64 
 11  perdida_capital    48842 non-null  int64 
 12  horas_por_semana   48842 non-null  int64 
 13  pais_origen        48842 non-null  object
 14  ingresos           48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


,0
edad,0
tipo_trabajo,0
peso_muestral,0
nivel_educativo,0
anios_educacion,0
estado_civil,0
ocupacion,0
relacion_familiar,0
raza,0
sexo,0


Se observa que de las 15 variables del dataset, el 40% son numéricas, y el 60% son categóricas. Adicionalmente, ninguna de las variables tiene valores nulos.

In [4]:
df.isin(["?"]).sum()

,0
edad,0
tipo_trabajo,2799
peso_muestral,0
nivel_educativo,0
anios_educacion,0
estado_civil,0
ocupacion,2809
relacion_familiar,0
raza,0
sexo,0


In [ ]:
df.replace("?", pd.NA, inplace=True)

Se identifican valores con el caracter "?" y se reemplazan por NA.

In [5]:
print('\nResumen categórico:')
display(df.describe(include='object'))


Resumen categórico:


,tipo_trabajo,nivel_educativo,estado_civil,ocupacion,relacion_familiar,raza,sexo,pais_origen,ingresos
count,48842,48842,48842,48842,48842,48842,48842,48842,48842
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,33906,15784,22379,6172,19716,41762,32650,43832,37155


Según el resumen categórico, se tienen 9 tipos de trabajo, 7 estados civiles, 15 ocupaciones, entre otros. En el top de tipo_trabajo resalta el Private, y la base tiene mas hombres que mujeres. Adicional, el país_origen que mas se presenta es Estados Unidos.

**Análisis descriptivo y resumenes pivotes**: Distribución de valores por columna

In [ ]:
for col in df.columns:
    print(f"\n🔹 {col}")
    print(df[col].value_counts())


🔹 edad
edad
36    1348
35    1337
33    1335
23    1329
31    1325
      ... 
88       6
85       5
87       3
89       2
86       1
Name: count, Length: 74, dtype: int64

🔹 tipo_trabajo
tipo_trabajo
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov            1981
Self-emp-inc         1695
Federal-gov          1432
Without-pay            21
Never-worked           10
Name: count, dtype: int64

🔹 peso_muestral
peso_muestral
203488    21
120277    19
190290    19
126569    18
125892    18
          ..
275677     1
311631     1
374763     1
242136     1
26711      1
Name: count, Length: 28523, dtype: int64

🔹 nivel_educativo
nivel_educativo
HS-grad         15784
Some-college    10878
Bachelors        8025
Masters          2657
Assoc-voc        2061
11th             1812
Assoc-acdm       1601
10th             1389
7th-8th           955
Prof-school       834
9th               756
12th              657
Doctorate         594
5th-6th           509
1st-4th

In [ ]:
cols = ["relacion_familiar", "raza", "sexo", "pais_origen"]

for col in cols:
    print(f"\n🔹 {col.upper()}")
    print(df[col].value_counts())


🔹 RELACION_FAMILIAR
relacion_familiar
Husband           19716
Not-in-family     12583
Own-child          7581
Unmarried          5125
Wife               2331
Other-relative     1506
Name: count, dtype: int64

🔹 RAZA
raza
White                 41762
Black                  4685
Asian-Pac-Islander     1519
Amer-Indian-Eskimo      470
Other                   406
Name: count, dtype: int64

🔹 SEXO
sexo
Male      32650
Female    16192
Name: count, dtype: int64

🔹 PAIS_ORIGEN
pais_origen
United-States                 43832
Mexico                          951
Philippines                     295
Germany                         206
Puerto-Rico                     184
Canada                          182
El-Salvador                     155
India                           151
Cuba                            138
England                         127
China                           122
South                           115
Jamaica                         106
Italy                           105
Dominican-

Proporciones % Genero, pais, sexo.

In [ ]:
for col in cols:
    print(f"\n🔹 {col.upper()} (%)")
    print((df[col].value_counts(normalize=True)*100).round(2))


🔹 RELACION_FAMILIAR (%)
relacion_familiar
Husband           40.37
Not-in-family     25.76
Own-child         15.52
Unmarried         10.49
Wife               4.77
Other-relative     3.08
Name: proportion, dtype: float64

🔹 RAZA (%)
raza
White                 85.50
Black                  9.59
Asian-Pac-Islander     3.11
Amer-Indian-Eskimo     0.96
Other                  0.83
Name: proportion, dtype: float64

🔹 SEXO (%)
sexo
Male      66.85
Female    33.15
Name: proportion, dtype: float64

🔹 PAIS_ORIGEN (%)
pais_origen
United-States                 91.35
Mexico                         1.98
Philippines                    0.61
Germany                        0.43
Puerto-Rico                    0.38
Canada                         0.38
El-Salvador                    0.32
India                          0.31
Cuba                           0.29
England                        0.26
China                          0.25
South                          0.24
Jamaica                        0.22
Italy    

In [ ]:
df["horas_por_semana"].value_counts().head(10)

,count
horas_por_semana,
40,22803
50,4246
45,2717
60,2177
35,1937
20,1862
30,1700
55,1051
25,958


Sexo vs ingresos

In [ ]:
pd.crosstab(df["sexo"], df["ingresos"], normalize="index")*100

ingresos,<=50K,>50K
sexo,,
Female,89.074852,10.925148
Male,69.623277,30.376723


Relación familiar vs ingresos

In [ ]:
pd.crosstab(df["relacion_familiar"], df["ingresos"], normalize="index")*100

ingresos,<=50K,>50K
relacion_familiar,,
Husband,55.132887,44.867113
Not-in-family,89.859334,10.140666
Other-relative,96.547145,3.452855
Own-child,98.535813,1.464187
Unmarried,93.970732,6.029268
Wife,53.110253,46.889747


Top Paises


In [ ]:
top_paises = df["pais_origen"].value_counts().head(5).index

pd.crosstab(df[df["pais_origen"].isin(top_paises)]["pais_origen"],
            df["ingresos"], normalize="index")*100

ingresos,<=50K,>50K
pais_origen,,
Germany,71.844660,28.155340
Mexico,95.057834,4.942166
Philippines,71.186441,28.813559
Puerto-Rico,89.130435,10.869565
United-States,75.602300,24.397700


“En promedio, quienes ganan más trabajan X horas vs Y horas”

In [ ]:
df.groupby("ingresos")["horas_por_semana"].mean()

,horas_por_semana
ingresos,
<=50K,38.840048
>50K,45.452896


“El análisis exploratorio muestra diferencias importantes en la distribución de ingresos según variables demográficas como género, relación familiar y país de origen, además de una relación con la cantidad de horas trabajadas.”

**Calidad de datos**: nulos, duplicados y algunas validaciones básicas

In [6]:
df.isna().sum()

,0
edad,0
tipo_trabajo,0
peso_muestral,0
nivel_educativo,0
anios_educacion,0
estado_civil,0
ocupacion,0
relacion_familiar,0
raza,0
sexo,0


In [7]:
df.nunique(dropna=False)

,0
edad,74
tipo_trabajo,9
peso_muestral,28523
nivel_educativo,16
anios_educacion,16
estado_civil,7
ocupacion,15
relacion_familiar,6
raza,5
sexo,2


In [15]:
quality = pd.DataFrame({
    'nulos': df.isna().sum(),
    'pct_nulos': (df.isna().mean() * 100).round(2),
    'n_unicos': df.nunique(dropna=False)
}).sort_values('pct_nulos', ascending=False)

display(quality)

print('Filas duplicadas:', df.duplicated().sum())
print('ganancia_capital negativos:', (df['ganancia_capital'] < 0).sum())
print('edad negativos:', (df['edad'] < 0).sum())
print('horas_por_semana negativos:', (df['horas_por_semana'] < 0).sum())

,nulos,pct_nulos,n_unicos
edad,0,0.0,74
tipo_trabajo,0,0.0,9
peso_muestral,0,0.0,28523
nivel_educativo,0,0.0,16
anios_educacion,0,0.0,16
estado_civil,0,0.0,7
ocupacion,0,0.0,15
relacion_familiar,0,0.0,6
raza,0,0.0,5
sexo,0,0.0,2


Filas duplicadas: 52
ganancia_capital negativos: 0
edad negativos: 0
horas_por_semana negativos: 0


Se identifica 52 filas duplicadas que se deben eliminar, y se comprueba que no hay valores negativos incongruentes.

**Limpieza de datos**: eliminación de filas; agrupación de categorías para mejor análisis.

In [17]:
df = df.drop_duplicates()
print('Shape final:', df.shape)
display(df.head())

Shape final: (48790, 15)


,edad,tipo_trabajo,peso_muestral,nivel_educativo,anios_educacion,estado_civil,ocupacion,relacion_familiar,raza,sexo,ganancia_capital,perdida_capital,horas_por_semana,pais_origen,ingresos
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


Se eliminan las 52 filas repetidas, quedando con un dataset de 48,790 registros

**Feature engineering**: crear nuevas variables para mejorar el análisis

In [34]:
bins = [17, 25, 35, 60, 99]
labels = ['Joven', 'Joven adulto', 'Adulto', 'Adulto mayor']
df["grupo_edad"] = pd.cut(
    df["edad"],
    bins=bins,
    labels=labels,
    right=True
)
df['grupo_edad'].value_counts()

,count
grupo_edad,
Adulto,22876
Joven adulto,12710
Joven,9005
Adulto mayor,3605


In [33]:
region_map = {
    # Norteamérica
    "United-States": "Norteamérica",
    "Canada": "Norteamérica",
    "Mexico": "Norteamérica",
    "Puerto-Rico": "Norteamérica",
    "Outlying-US(Guam-USVI-etc)": "Norteamérica",

    # Latinoamérica y Caribe
    "El-Salvador": "Latinoamérica",
    "Cuba": "Latinoamérica",
    "Jamaica": "Latinoamérica",
    "Dominican-Republic": "Latinoamérica",
    "Guatemala": "Latinoamérica",
    "Haiti": "Latinoamérica",
    "Nicaragua": "Latinoamérica",
    "Peru": "Latinoamérica",
    "Ecuador": "Latinoamérica",
    "Columbia": "Latinoamérica",
    "Honduras": "Latinoamérica",
    "Trinadad&Tobago": "Latinoamérica",

    # Europa
    "Germany": "Europa",
    "England": "Europa",
    "Italy": "Europa",
    "Poland": "Europa",
    "Portugal": "Europa",
    "Greece": "Europa",
    "France": "Europa",
    "Ireland": "Europa",
    "Scotland": "Europa",
    "Hungary": "Europa",
    "Yugoslavia": "Europa",
    "Holand-Netherlands": "Europa",

    # Asia
    "Philippines": "Asia",
    "India": "Asia",
    "China": "Asia",
    "Japan": "Asia",
    "Vietnam": "Asia",
    "Taiwan": "Asia",
    "Iran": "Asia",
    "Thailand": "Asia",
    "Hong": "Asia",
    "Cambodia": "Asia",
    "Laos": "Asia",

    # Otros
    "South": "Otros"
}
df["region"] = df["pais_origen"].map(region_map)
df['region'].value_counts()

,count
region,
Norteamérica,45124
Asia,980
Latinoamérica,935
Europa,780
Otros,115


In [35]:
df

,edad,tipo_trabajo,peso_muestral,nivel_educativo,anios_educacion,estado_civil,ocupacion,relacion_familiar,raza,sexo,ganancia_capital,perdida_capital,horas_por_semana,pais_origen,ingresos,grupo_edad,region
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K,Adulto,Norteamérica
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K,Adulto,Norteamérica
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K,Adulto,Norteamérica
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K,Adulto,Norteamérica
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K,Joven adulto,Latinoamérica
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
48837,39,Private,215419,Bachelors,13,Divorced,Prof-specialty,Not-in-family,White,Female,0,0,36,United-States,<=50K,Adulto,Norteamérica
48838,64,?,321403,HS-grad,9,Widowed,?,Other-relative,Black,Male,0,0,40,United-States,<=50K,Adulto mayor,Norteamérica
48839,38,Private,374983,Bachelors,13,Married-civ-spouse,Prof-specialty,Husband,White,Male,0,0,50,United-States,<=50K,Adulto,Norteamérica
48840,44,Private,83891,Bachelors,13,Divorced,Adm-clerical,Own-child,Asian-Pac-Islander,Male,5455,0,40,United-States,<=50K,Adulto,Norteamérica


Dado que las variables edad y pais_origen tienen demasiados valores diferentes se decide clusterizarlas, creando dos variables nuevas: rango_edad y region, tal como se muestra en las dos últimas columnas de la visualización anterior del dataset, quedando con 17 columnas.